### PR TITLE
[TensorOp] Add Tensor Op class

### PR DESCRIPTION
**Pull request by kparichay**
Add tensor op class
Also update FC layer as per tensor op class
Also add graph class for tensor op

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: nntrainer/include/graph.h [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+   */
+  void registerTensorOp(TensorOp op) {}
+
+  /**
+   * @brief Get the all the tensors object
+   *
+   * @return std::vector<Tensor>& List of all the tensor objects for the all
+   * registered ops
+   */
+  std::vector<Tensor> &getTensors() { return all_tensors; }
```
Comment (zhoonit) [inline nntrainer/include/graph.h:47]: Q. So, is `std::vector<Tensor>` :cactus: gonna be changed to a map to save inputs and outputs?
I am confused if :cactus: gonna be a storage(for inputs/outputs) or graph or both?

Comment (kparichay) [reply nntrainer/include/graph.h:47]: In a graph, we have a list of vectors and a list of edges connecting them.
Here, std::vector<Tensor> is the information on each edge.

File: nntrainer/src/fc_layer_op.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  dim.batch(1);
+  Tensor weight = initializeWeight(dim, weight_ini_type, status);
+  NN_RETURN_STATUS();
+
+  setParamSize(2);
+  paramsAt(0) = {std::move(weight), Tensor(weight.getDim()), "FC:weight"};
+  paramsAt(1) = {std::move(bias), Tensor(bias.getDim()), "FC:bias"};
+
+  TensorParam w =
+    TensorParam(paramsAt(0).weight, Tensor(weight.getDim()), "FC:weight", true);
```
Comment (jijoongmoon) [inline nntrainer/src/fc_layer_op.cpp:63]: may be ```TensorParam(paramsAt(0).weight, paramsAt(0).grad, "FC:weight"}``` and TensorParam(paramsAt(1).weight, paramsAt(1).grad,"FC:bias",true)```

File: nntrainer/src/fc_layer_op.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+           .run()
+           .sum(0);
+}
+
+void FullyConnectedLayer::applyGrad(int iteration) {
+  int idx = 0;
+  for (auto &tp : inputs_op) {
+    if (tp.trainable && trainable) {
+      opt.apply_gradients(
+        std::make_shared<UpdatableParam>(params.get()[idx - 1]), 1, iteration);
```
Comment (jijoongmoon) [inline nntrainer/src/fc_layer_op.cpp:109]: Do we need updatableParam cause we have tensorparam.

File: nntrainer/src/neuralnet.cpp [hunk 4]
```diff
@@ -824,6 +824,24 @@ int NeuralNetwork::addLayer(std::shared_ptr<Layer> layer) {
...
+  if (type == LAYER_UNKNOWN)
+    return ML_ERROR_INVALID_PARAMETER;
+
+  if (initialized) {
+    return ML_ERROR_NOT_SUPPORTED;
+  }
+
+  ensureName(layer);
+  std::shared_ptr<TensorOp> op = std::dynamic_pointer_cast<TensorOp>(layer);
+  graph.registerTensorOp(*op);
```
Comment (jijoongmoon) [inline nntrainer/src/neuralnet.cpp:840]: I'm not sure if tensorop could be tensor calculation level not a layer level. Cause there are calculation with tensor values like kernel calculation or make proper data type in convolution.

File: nntrainer/src/fc_layer_op.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/src/fc_layer_op.cpp:109]: Yes, they will be merged. This PR is just to demonstrate a sample example.

File: nntrainer/src/neuralnet.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/src/neuralnet.cpp:840]: As you said, custom op for tensor is not good.
As you might have noticed, `tensor_op`'s method `compute_op` and other methods do not take any arguments.
The reason is to allow multiple arguments.

When tensorOp does tensor level calculation later, lazy tensor would have to be deprecated as tensor would do all the lazily itself.

Please tell me if I misunderstood your question.

**Comment by kparichay** [comment]
Note: this setup has an issue where layer configuration cannot be passed to the `tensor_op` while computing.
For example - in convolution `forwarding`, various configuration values such as stride, padding etc needs to be accessible to tensor_op so as to perform this function (this problem is only for layers. Tensor calculation does not need such configurations). Two possible solutions - 
1. TensorOp has properties than layers and then it can use it easily
2. These properties be passed to the computation functions (forwarding/backwarding or computeop/computegrad). However, in this case, graph needs to store these configurations for these tensorOp (essentially making it same as above).

Further, with operations such as gemm etc (which don't operate at tensor level), must think about how they will be covered with tensorOp.

File: nntrainer/src/fc_layer_op.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/src/fc_layer_op.cpp:63]: yes, that is better. Will be updated to this style.

**Comment by zhoonit** [comment]
> Note: this setup has an issue where layer configuration cannot be passed to the `tensor_op` while computing.
> For example - in convolution `forwarding`, various configuration values such as stride, padding etc needs to be accessible to tensor_op so as to perform this function (this problem is only for layers. Tensor calculation does not need such configurations). Two possible solutions -
> 
> 1. TensorOp has properties than layers and then it can use it easily
> 2. These properties be passed to the computation functions (forwarding/backwarding or computeop/computegrad). However, in this case, graph needs to store these configurations for these tensorOp (essentially making it same as above).
> 
> Further, with operations such as gemm etc (which don't operate at tensor level), must think about how they will be covered with tensorOp.

I think..
1. Tensor Ops should have inner state by inheriting base class(TensorOps)
2. Gemm should be interfaced by tensor level.

**Comment by kparichay** [comment]
> I think..
> 
> 1. Tensor Ops should have inner state by inheriting base class(TensorOps)
> 2. Gemm should be interfaced by tensor level wrapper operation

For 2. , yes that sounds good. If any bigger operation wants to use gemm, then that operation should becomes a tesnorOp itself.
However for 1, there comes a requirement for a structure to hold these properties of various data types and what not (maybe std::map(std::string, std::any) might satisfy).

**Comment by kparichay** [comment]
Closing as this isn't needed now.

**Pull request closed without merging**